pkg/kubelet/kubeletconfig/util/codec/codec.go

```diff
@@ -78,10 +81,35 @@ func NewYAMLEncoder(groupName string) (runtime.Encoder, error) {
 
 // DecodeKubeletConfiguration decodes a serialized KubeletConfiguration to the internal type
 func DecodeKubeletConfiguration(kubeletCodecs *serializer.CodecFactory, data []byte) (*kubeletconfig.KubeletConfiguration, error) {
+	var (
+		obj runtime.Object
+		gvk *schema.GroupVersionKind
+	)
+
 	// the UniversalDecoder runs defaulting and returns the internal type by default
 	obj, gvk, err := kubeletCodecs.UniversalDecoder().Decode(data, nil, nil)
 	if err != nil {
-		return nil, fmt.Errorf("failed to decode, error: %v", err)
+		// For now allow lenient decoding without strict serializer. This
+		// path is to be dropped when v1beta1 is deprecated and removed.
+		if runtime.IsStrictDecodingError(err) {
```

Comment:
Rather than constructing the non-strict codec here, let's change DecodeKubeletConfiguration to take a strict and legacy non-strict codec as arguments. Then, NewSchemeAndCodecs can return both the strict and non-strict codecs. I think this is what @liggitt meant.

Reply:
Ok got it. I just want to double check since changing the return values on `NewSchemeAndCodecs` would require me to touch a couple of files: 
- [cmd/kubelet/app/server.go](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L430)
- [cmd/kubelet/app/options/options.go](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/options/options.go#L269)
- [pkg/kubelet/apis/config/scheme/scheme_test.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/scheme/scheme_test.go)
- [pkg/kubelet/kubeletconfig/checkpoint/download.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubeletconfig/checkpoint/download.go#L110)
- [pkg/kubelet/kubeletconfig/checkpoint/store/fsstore_test.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubeletconfig/checkpoint/store/fsstore_test.go#L705)
- [pkg/kubelet/kubeletconfig/configfiles/configfiles.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubeletconfig/configfiles/configfiles.go#L48)

In those places where the returned codec is important, I'm assuming to use the strict codecs that would be returned by `NewSchemeAndCodecs`?

Reply:
If it is fewer ripples to build a lenient one in place (like the PR currently does), and that is the only place we try to decode a kubelet config, I'm fine with the current approach.

Reply:
@liggitt I think there are two places: when decoding a [local](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubeletconfig/util/codec/codec.go#L80) and [remote](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubeletconfig/checkpoint/download.go#L108) config.